### PR TITLE
Fix Python 3 errors

### PR DESCRIPTION
**Pull request by davedoesdev**
- Signing data should be byte strings
- Don't log exception outside handler (#300)

### CODE REVIEW

File: tuf/pyca_crypto_keys.py [hunk 1]
```diff
@@ -428,7 +428,7 @@ def verify_rsa_signature(signature, signature_method, public_key, data):
                                   salt_length=hashes.SHA256().digest_size), 
                                   hashes.SHA256())
 
-      verifier.update(data.encode('utf-8'))
+      verifier.update(data)
```
Comment (vladimir-v-diaz) [inline tuf/pyca_crypto_keys.py]: The corresponding functions for `pycrypto_keys.py`will also need this change.

https://github.com/theupdateframework/tuf/blob/develop/tuf/pycrypto_keys.py#L297
https://github.com/theupdateframework/tuf/blob/develop/tuf/pycrypto_keys.py#L399

You can verify the required changes by running `test_pycrypto_keys.py`.
See the Travis build failure here: https://travis-ci.org/theupdateframework/tuf#L1294

File: tuf/client/updater.py [hunk 2]
```diff
@@ -1226,8 +1226,8 @@ def _get_file(self, filepath, verify_file_function, file_type,
       return file_object
     
     else:
-      logger.exception('Failed to update {0} from all mirrors: {1}'.format(
```
Comment (vladimir-v-diaz) [inline tuf/client/updater.py]: Which version of Python 3 are you testing?  I can't reproduce this error.

File: tuf/pyca_crypto_keys.py [hunk 3]
```diff
@@ -404,7 +404,7 @@ def verify_rsa_signature(signature, signature_method, public_key, data):
   tuf.formats.NAME_SCHEMA.check_match(signature_method)
 
   # Does 'signature' have the correct format?
-  tuf.formats.PYCRYPTOSIGNATURE_SCHEMA.check_match(signature)
+  tuf.formats.PYCACRYPTOSIGNATURE_SCHEMA.check_match(signature)
```
Comment (vladimir-v-diaz) [inline tuf/pyca_crypto_keys.py]: Following this change, there are two test cases that need to verify that a `tuf.FormatError` exception is raised.

1. https://github.com/theupdateframework/tuf/blob/develop/tests/test_pyca_crypto_keys.py#L83-L84

2. [Line 87](https://github.com/theupdateframework/tuf/blob/develop/tests/test_pycrypto_keys.py#L87) of `test_pycrypto_keys.py` can be changed to:
```Python
self.assertRaises(tuf.FormatError, pycrypto.create_rsa_signature, private_rsa, '')
```

**Comment by vladimir-v-diaz** [comment]
Thanks for the pull request!

This pull request is ready to be merged once the test cases indicated in the in-line comments are updated.

**Comment by davedoesdev** [comment]
OK I made those changes.
Regarding the `logger.exception` issue, you get it on Python 3 if you enable console messages, i.e.:

```python
diff --git a/tests/test_updater.py b/tests/test_updater.py
index 40b005b..25e3198 100755
--- a/tests/test_updater.py
+++ b/tests/test_updater.py
@@ -76,7 +76,7 @@ import tuf.client.updater as updater
 import six
 
 logger = logging.getLogger('tuf.test_updater')
-repo_tool.disable_console_log_messages()
+#repo_tool.disable_console_log_messages()
 
 
```

**Comment by vladimir-v-diaz** [comment]
Thanks for helping me reproduce the error.  I've now verified that [line 1229](https://github.com/theupdateframework/tuf/pull/301#discussion-diff-49613619L1229) of updater.py does indeed raise an error if console logging is enabled.

I searched for all other instance of logger.exception, and line 1229 is the only place where it's not called from an exception handler.

All the unit tests are passing now (apart of an unrelated test case failure) so I'll go ahead and merge.

**Comment by vladimir-v-diaz** [comment]
@davedoesdev We might need to fix a few other similar Python 3 issues with the PyCA Cryptography module (or it might be because I upgraded my local version)

**Comment by davedoesdev** [comment]
Yes, this is what I get:

```
======================================================================
ERROR: test_generate_and_write_ed25519_keypair (test_repository_lib.TestRepositoryToolFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/tuf/tests/test_repository_lib.py", line 214, in test_generate_and_write_ed25519_keypair
    repo_lib.generate_and_write_ed25519_keypair(test_keypath, password='pw')
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/repository_lib.py", line 988, in generate_and_write_ed25519_keypair
    encrypted_key = tuf.keys.encrypt_key(ed25519_key, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 1231, in encrypt_key
    tuf.pyca_crypto_keys.encrypt_key(key_object, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 750, in encrypt_key
    encrypted_key = _encrypt(json.dumps(key_object), derived_key_information)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 920, in _encrypt
    ciphertext = encryptor.update(key_data) + encryptor.finalize()
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 137, in update
    return self._ctx.update(data)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 117, in update
    len(data))
TypeError: initializer for ctype 'unsigned char *' must be a bytes or list or tuple, not str

======================================================================
ERROR: test_import_ed25519_privatekey_from_file (test_repository_lib.TestRepositoryToolFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/tuf/tests/test_repository_lib.py", line 290, in test_import_ed25519_privatekey_from_file
    repo_lib.generate_and_write_ed25519_keypair(ed25519_keypath, password='pw')
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/repository_lib.py", line 988, in generate_and_write_ed25519_keypair
    encrypted_key = tuf.keys.encrypt_key(ed25519_key, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 1231, in encrypt_key
    tuf.pyca_crypto_keys.encrypt_key(key_object, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 750, in encrypt_key
    encrypted_key = _encrypt(json.dumps(key_object), derived_key_information)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 920, in _encrypt
    ciphertext = encryptor.update(key_data) + encryptor.finalize()
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 137, in update
    return self._ctx.update(data)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 117, in update
    len(data))
TypeError: initializer for ctype 'unsigned char *' must be a bytes or list or tuple, not str

======================================================================
ERROR: test_import_ed25519_publickey_from_file (test_repository_lib.TestRepositoryToolFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/tuf/tests/test_repository_lib.py", line 242, in test_import_ed25519_publickey_from_file
    repo_lib.generate_and_write_ed25519_keypair(ed25519_keypath, password='pw')
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/repository_lib.py", line 988, in generate_and_write_ed25519_keypair
    encrypted_key = tuf.keys.encrypt_key(ed25519_key, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 1231, in encrypt_key
    tuf.pyca_crypto_keys.encrypt_key(key_object, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 750, in encrypt_key
    encrypted_key = _encrypt(json.dumps(key_object), derived_key_information)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 920, in _encrypt
    ciphertext = encryptor.update(key_data) + encryptor.finalize()
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 137, in update
    return self._ctx.update(data)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 117, in update
    len(data))
TypeError: initializer for ctype 'unsigned char *' must be a bytes or list or tuple, not str

======================================================================
ERROR: test_decrypt_key (test_keys.TestKeys)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/tuf/tests/test_keys.py", line 277, in test_decrypt_key
    encrypted_key = KEYS.encrypt_key(self.rsakey_dict, passphrase).encode('utf-8')
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 1231, in encrypt_key
    tuf.pyca_crypto_keys.encrypt_key(key_object, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 750, in encrypt_key
    encrypted_key = _encrypt(json.dumps(key_object), derived_key_information)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/pyca_crypto_keys.py", line 920, in _encrypt
    ciphertext = encryptor.update(key_data) + encryptor.finalize()
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 137, in update
    return self._ctx.update(data)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 117, in update
    len(data))
TypeError: initializer for ctype 'unsigned char *' must be a bytes or list or tuple, not str

======================================================================
ERROR: test_https_connection (test_download.TestDownload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.4/urllib/request.py", line 1183, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/usr/lib/python3.4/http/client.py", line 1137, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.4/http/client.py", line 1182, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.4/http/client.py", line 1133, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.4/http/client.py", line 963, in _send_output
    self.send(msg)
  File "/usr/lib/python3.4/http/client.py", line 898, in send
    self.connect()
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/download.py", line 674, in connect
    ca_certs=cert_path)
  File "/usr/lib/python3.4/ssl.py", line 890, in wrap_socket
    ciphers=ciphers)
  File "/usr/lib/python3.4/ssl.py", line 580, in __init__
    self.do_handshake()
  File "/usr/lib/python3.4/ssl.py", line 807, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/david/tuf/tests/test_download.py", line 236, in test_https_connection
    download.safe_download(https_url, target_data_length)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/download.py", line 115, in safe_download
    return _download_file(url, required_length, STRICT_REQUIRED_LENGTH=True)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/download.py", line 241, in _download_file
    connection = _open_connection(url)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/download.py", line 468, in _open_connection
    return opener.open(request, timeout = tuf.conf.SOCKET_TIMEOUT)
  File "/usr/lib/python3.4/urllib/request.py", line 464, in open
    response = self._open(req, data)
  File "/usr/lib/python3.4/urllib/request.py", line 482, in _open
    '_open', req)
  File "/usr/lib/python3.4/urllib/request.py", line 442, in _call_chain
    result = func(*args)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/download.py", line 694, in https_open
    return self.do_open(self.specialized_conn_class, req)
  File "/usr/lib/python3.4/urllib/request.py", line 1185, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)>

======================================================================
ERROR: test_with_tuf (test_indefinite_freeze_attack.TestIndefiniteFreezeAttack)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/tuf/tests/test_indefinite_freeze_attack.py", line 236, in test_with_tuf
    'password')
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/repository_lib.py", line 870, in import_rsa_privatekey_from_file
    rsa_key = tuf.keys.import_rsakey_from_encrypted_pem(encrypted_pem, password)
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 958, in import_rsakey_from_encrypted_pem
    check_crypto_libraries(['rsa', 'general'])
  File "/home/david/tuf/.tox/py34/lib/python3.4/site-packages/tuf/keys.py", line 602, in check_crypto_libraries
    raise tuf.UnsupportedLibraryError(message)
tuf.UnsupportedLibraryError: The 'pyca-cryptography' crypto library specified in "tuf.conf.GENERAL_CRYPTO_LIBRARY" could not be imported.

----------------------------------------------------------------------
Ran 242 tests in 174.946s

FAILED (errors=6)
```

The certificate verification happens on 2.7 as well but the other ones need fixing on 3.4

**Comment by davedoesdev** [comment]
https://github.com/theupdateframework/tuf/pull/305

**Pull request merged**